src/coordinator/mod.rs

```diff
@@ -409,6 +413,52 @@ impl Coordinator {
         Ok(())
     }
 
+    // Unlike input registers, holding registers are not broadcast by inverters,
+    // but they are interesting nevertheless. Broadcasting the holding registers
+    // when we connect to an inverter makes it easy for configuration data to be
+    // tracked, which is particularly useful in conjunction with HomeAssistant.
+    async fn inverter_connected(&self, datalog: Serial) -> Result<()> {
+        let inverter = match self.config.enabled_inverter_with_datalog(datalog) {
+            Some(inverter) => inverter,
+            None => bail!("Unknown inverter connected: {}", datalog),
+        };
+
+        if !inverter.broadcast_holding_registers() {
+            return Ok(());
+        }
+
+        info!("Reading holding registers for inverter {}", datalog);
+
+        // We can only read holding registers in blocks of 40. Reading 41 bytes
```

Comment:
Discovered by experimentation. Here's a marked-up, anonymised version of the packet I get when I set count to 41:

<details>

```
[ // TCP header
  0xa1, 0x1a, // prefix
  2, 0, // protocol version
  31, 0, // packet length
  1, // address
  0xc4, // tcp function
  66, 65, 49, 49, 49, 49, 49, 49, 49, 49, // datalog SN
  
  // TranslatedData
  17, // address?
  0, // device function
  1, 131, // ?
  50, 50, 50, 50, 50, 50, 50, 50, 50, 50, // inverter SN
  0, 0, // length?
  3, 81, 84 // ?
]
```

</details>

Presumably this expresses "error" somehow, and we could parse it and return a polite error out of `read_hold` instead of the current behaviour, which is to kill the TCP connection.

Alternatively, maybe it expresses a different type of response that is only used when > 40 bytes is requested in a holding register, and all the data is in there *somewhere* :sweat_smile:

Reply:
Yeah this is a limitation of these inverters. I *thought* I had documented this in the wiki but I can't find it now so maybe I didn't.

Basically registers 0-39 have to be read at once, not overlapping past 39, then similarly for 40-79, 80-119 and so on. I've seen this mentioned in a modbus protocol documentation for the inverter. I'll try and squeeze it into the wiki documentation somewhere.

The 131 in your details above correlates to ReadHoldError: https://github.com/celsworth/lxp-bridge/blob/master/src/lxp/packet.rs#L519

Not sure what the details of the payload are exactly in this case. The final two ints will just be checksum, so you're basically getting "3" back in your response I think.. whatever that means :)

Reply:
Aha. So this is basically `ReadHold1`, `ReadHold2`, etc, in reality?

Now that we're reading all the holding registers (that we know about), I wonder if it's worth following the example of `ReadInput` and publishing a consolidated message with everything in it? Having the in-memory map of the whole set would solve the issue down in https://github.com/celsworth/lxp-bridge/pull/147#discussion_r1111125255, for instance, and doing all the interpretation server-side would make the messages more useful generally.

Reply:
`ReadInput1`, `ReadInput2` etc, yeah. Although there's a but, of course - some newer inverters/firmwares seem to be able to send 0-127 at once. Not sure how to determine which ones those are unfortunately..

I think if we're adding a consolidated message with all holds we need to be careful not to design ourselves into a confusing hole. So for input registers, the MQTT topics `read/inputs/1` (plural inputs) is *parsed* values, ie whats in a register as decivolt is normalised into volts; and 1 isn't the register number, its the first *set* of registers. Whereas `read/input/1` is a raw value of input register 1.

I tried documenting/explaining at https://github.com/celsworth/lxp-bridge/wiki/MQTT-Publishes but it probably made a lot more sense in my head than it ever makes to anyone else ;)

So I'd like to keep the same system for holding registers as well.. `read/hold/1` is a raw register 1`, `read/holds/all` is parsed registers. This is probably not trivial though. We'd need to know how to normalise each holding register for starters. And then devise a way of doing it; we got it cheaply for input registers via nom parsing.

Reply:
Aha, it's all making a lot more sense now :thumbsup:. I wonder if we can see a future where we don't publish individual holding registers either.

In this PR, I think I'd be OK with continuing to publish the individual ones and not making much headway towards `read/holds/all`. We can always change the HA audiscovery messages to refer to the combined one once we get it in, after all.

As for packet, if we're just wrapping modbus with this protocol, maybe https://www.simplymodbus.ca/exceptions.htm is a reasonable guide - in which case, 0x03 would be "Illegal Data Value", i.e., bad request.

It doesn't super-hurt to cycle the TCP connection, but it would be neater to correctly parse it and recognise that it's safe to re-use. Maybe there's a general rust modbus library that could replace all the custom stuff? Probs one for an issue :sweat_smile:

Reply:
Not publishing individual holding registers would be quite a leap and change in functionality. Not sure how many people really rely on that, I know I quite like poking at individual registers to change things like battery charge voltage quite often.

This would need some careful designing for a useful replacement, if we read them all on connection then update our local copy whenever they change (in the case of WriteMulti, we see what changed but not what to; we can just read those registers again to update them), then publish an `all` style JSON dump on demand, or at intervals, as per configuration? Or when any of them change? Could be quite interesting. But any further discussion here should probably be in its own issue or its going to get lost :)

Dropping the TCP connection on error is just an abundance of caution really, I took the view that if we don't know whats going on we'll just drop it and start over. Certainly not opposed to adding better error handling here, especially when the error is caused by a malformed request we sent or something.